### ISSUE TITLE
Direct database import does not bring alternatives

### ISSUE THREAD

**Issue by ghost** [Spine-project/Spine-Toolbox#1613]
Importing from another datastore directly via executing a datastore does not import alternatives. The import log show these errors:

```
Alternative 'base' not found.
Alternative 'low-price' not found.
Alternative 'thermal-generation' not found.
Alternative 'low-sector-coupling' not found.
Alternative 'low-electrification' not found.
Alternative 'little-flexible-demand' not found.
Could not import parameter value for 'LNG', class 'node', parameter 'price': alternative 'med-price' does not exist.
Could not import parameter value for 'biomethane', class 'node', parameter 'price': alternative 'high-price' does not exist.
Could not import parameter value for 'biomethane', class 'node', parameter 'price': alternative 'low-price' does not exist.
Could not import parameter value for 'coal', class 'node', parameter 'price': alternative 'med-price' does not exist.
Could not import parameter value for 'hvyoil', class 'node', parameter 'price': alternative 'med-price' does not exist.
Could not import parameter value for 'hydrogen', class 'node', parameter 'price': alternative 'med-price' does not exist.
Could not import parameter value for 'lgtoil', class 'node', parameter 'price': alternative 'med-price' does not exist.
Could not import parameter value for 'lignite', class 'node', parameter 'price': alternative 'med-price' does not exist.
Could not import parameter value for 'lqdbiofuels', class 'node', parameter 'price': alternative 'high-price' does not exist.
Could not import parameter value for 'lqdbiofuels', class 'node', parameter 'price': alternative 'low-price' does not exist.
Could not import parameter value for 'natgas', class 'node', parameter 'price': alternative 'med-price' does not exist.
Could not import parameter value for 'nuclear', class 'node', parameter 'price': alternative 'med-price' does not exist.
Could not import parameter value for 'oilshale', class 'node', parameter 'price': alternative 'med-price' does not exist.
Could not import parameter value for 'procres', class 'node', parameter 'price': alternative 'high-price' does not exist.
Could not import parameter value for 'procres', class 'node', parameter 'price': alternative 'low-price' does not exist.
Could not import parameter value for 'residues', class 'node', parameter 'price': alternative 'high-price' does not exist.
Could not import parameter value for 'residues', class 'node', parameter 'price': alternative 'low-price' does not exist.
Could not import parameter value for 'wood', class 'node', parameter 'price': alternative 'high-price' does not exist.
Could not import parameter value for 'wood', class 'node', parameter 'price': alternative 'low-price' does not exist.
```

I wonder why the alternatives are not considered in the import.

**Comment by ererkka** [comment]
Same applies to Merger now.

**Comment by manuelma** [comment]
I couldn't reproduce this error, see screencast below. What am I missing?


https://user-images.githubusercontent.com/11472756/150095820-6314c53b-3f35-4964-bdff-907b2c648127.mp4

**Comment by soininen** [comment]
I believe I too saw this issue a month or two ago but now I cannot reprocude it anymore. Maybe it got accidentally fixed?

**Comment by manuelma** [comment]
Let's close, please anybody reopen if you happen to see it.

**Comment by nhniina** [comment]
I cannot re-open this issue - maybe I'm not a collaborator - but the issue is still there when merging to a mysql database. Merging to an sqlite works well.

**Comment by soininen** [comment]
I just tried out merging parameter values with alternatives on a MySQL database and didn't see anything out of ordinary. The target MySQL database had the class, the object, the parameter, the alternative and the value after the merge, no errors. Could this be something very specific to the TradeRES database? Or is it something that happens only sporadically or at specific conditions?

**Comment by soininen** [comment]
I can actually reproduce this on the TradeRES database.

**Comment by soininen** [comment]
Here is what I think goes wrong in this case:

- The target database has been configured to be case *insensitive* when it comes to unique values in columns, like alternative names
- In the target database the 'Base' alternative has been replaced by another alternative called 'base' (with id different from 1)
- On the other hand, the source database has the 'Base' alternative

What happens when Merger executes:
1. It tries to import alternatives from source database, including 'Base'
2. `spinedb_api` is case *sensitive* when it checks that alternative's name is unique => 'Base' needs to be imported into target
3. Target database refuses to add 'Base' as it conflicts with 'base'
4. `DatabaseMappingAddMixin._do_add_items()` raises an exception skipping the addition of the remaining alternatives

Possible workarounds to this would be to

- rename 'base' to 'Base' in the target database

or

- remove 'Base' from the source database; this might be troublesome if there are parameter values with the alterantive

The error message from `_do_add_items()` gets lost in the way and we should at least fix that. What to do with the case sensitivity issue? @manuelma?

**Comment by jkiviluo** [comment]
As a generic principle, I think one shouldn't distinguish data by using case differences in entity or alternative names. That's just dangerous.

How should we treat that when encountered? I think nicest would be to give a warning but allow the user to choose which form of the word to keep using (with a possibility to remember that choice). That's probably too heavy to implement, so, IMO, the 2nd best is to use the target database form (and leave a note in the log - but not for every instance, just once per import is enough).

**Comment by soininen** [comment]
> That's probably too heavy to implement, so, IMO, the 2nd best is to use the target database form (and leave a note in the log - but not for every instance, just once per import is enough).

Sounds like we should just declare names case-insentivie and call it a day. No need for warnings even.

**Comment by manuelma** [comment]
I feel we should discuss this. Case insensitivy is really unelegant imo.

**Comment by manuelma** [comment]
If I understand correctly, the SQL engine is case insensitive? Which dialect is this? I think we can try and adapt our API check to the underlying engine in terms of which case to use.

**Comment by soininen** [comment]
> If I understand correctly, the SQL engine is case insensitive?

It is MySQL and if I understood correctly, it is case insensitive by default.

How would we deal with conflicts that raise when merging data from case-sensitive database to a case-insensitive one? Is it possible to force the database engines to the same sensitivity at creation? If not, I'd seriously consider going full case-insensitive to avoid all these problems.

**Comment by manuelma** [comment]
Thank you @soininen I understand now.

I think we can forget about the origin of the data and just look at the target database. Is the underlying dialect case-insensitive? If yes, then `check_functions` should consider this fact when checking uniqueness. In the case of 'base' being added to a DB that has 'Base', then the check in `check_functions` should complain and say that it already exists.

**Comment by soininen** [comment]
> In the case of 'base' being added to a DB that has 'Base', then the check in check_functions should complain and say that it already exists.

I think the preferred action would be to *ignore* or update the item if it exists in target database instead of complaining. That's what we normally do, right? Additionally, in case of alternatives, we'd need to import parameter values so that they'd be added to the 'base' alternative replacing 'Base'. Similar care would be needed for scenario alternatives and perhaps others. Sounds like we'd need a switch to some kind of general case-insensitivity mode.

**Comment by manuelma** [comment]
> I think the preferred action would be to ignore or update the item if it exists in target database instead of complaining. That's what we normally do, right?

That's true, but we do that in the import after the check function has complained if I'm not mistaken. So the preferred action for the check function is to complain because the alternative exists, and let the import take the decision to update.

> Sounds like we'd need a switch to some kind of general case-insensitivity mode.

Yes, we would need to be able to lookup names from queries using the correct case-sensitivity. That can be a big effort and needs to be weigthed against the drawback of going full case-insensitive.

My preference I feel is to stay case-sensitive when possible, for readability and elegance mostly.

**Comment by soininen** [comment]
> That's true, but we do that in the import after the check function has complained if I'm not mistaken. So the preferred action for the check function is to complain because the alternative exists, and let the import take the decision to update.

True, thanks for pointing that out.

I'll improve error reporting a bit and then move on to other things as we now know how to work around the issue here.

**Comment by manuelma** [comment]
Sounds good, nice stuff @soininen
